buildarr/buildarr
# Add unit/functional tests and coverage checks

Testing related changes:

* Add unit tests for particular classes used by plugins (e.g. configuration model base class, attribute classes).
* Add functional tests for running Buildarr end-to-end when using it for a variety of scenarios.
* Update the `dummy` plugin, and add a new `dummy2` plugin, for use in the functional tests (hidden when not in testing mode).
* Add JUnit unit test metadata generation.
* Add Coverage.py coverage metadata generation (current result is 91% coverage).
* Update the GitHub Actions test workflow to run tests against both Ubuntu and Windows, using all supported Python versions (3.8-3.12), and report pass-fail/coverage results.
* Update all other GitHub Actions jobs to run using Python 3.12.

As a result of the testing added by this PR, fix a variety of bugs, and make a number of improvements:

* Fix Python 3.12 support by replacing usage of the removed `distutils.util.strtobool` function.
* Fix Windows support for the `buildarr compose` command, by converting native Windows paths returned by plugins to Unix paths suitable for running in Docker Desktop.
* Instead of a `str`, accept a 2-tuple or 3-tuple for representing short-form Docker volume syntax (splitting Windows paths using the `str`-based format was too troublesome).
* Add support for handling the `SIGBREAK` signal on Windows.
* Fix capitalisation on the daemon `--update-day` valid values list.
* Improve error message when an invalid daemon `--update-day` value is supplied.
* Make the `buildarr.trash_metadata_dir_prefix` configuration attribute nullable, to improve support for custom TRaSH metadata URLs.
* To make operations against all plugins deterministic, change `state.active_plugin` to a sorted `tuple` instead of a `frozenset`, and update a couple of loops to iterate over sorted lists of elements.
* Implement the `url_base` attribute in the `ConfigBase` base class, so plugins do not have to implement it separately.
    * Plugins should remove their separate implementation of this field when upgrading to the new version.
* Remove some redundant code.
* Fix an issue where `TrashID` fields will allow invalid characters to be defined in configuration files.

Other changes:

* Update copyright on all files.
* Remove some now-redundant linter silencing.
* Fix some minor metadata errors in `pyproject.toml`.
* Upgrade PDM to 2.15.0.
